src/Toml/Codec/Combinators.hs

```diff
@@ -517,33 +518,86 @@ map keyCodec valCodec key = Codec input output
-    -> TomlBiMap val AnyValue  -- ^ Bidirectional converter between TOML and Map values
-    -> Key  -- ^ Table name key
-    -> TomlCodec (Map key val)
+tableMapCodec :: forall key val .
+                 Ord key
+              => TomlBiMap Key key  -- ^ Bidirectional converter between TOML and Map keys
+              -> TomlBiMap val AnyValue  -- ^ Bidirectional converter between TOML and Map values
+              -> Key  -- ^ Table name key
+              -> TomlCodec (Map key val)
 tableMapCodec keyBiMap valBiMap key = Codec input output
   where
     input :: TomlEnv (Map key val)
     input = do
-        mTable <- asks $ Prefix.lookup key . tomlTables
-        case mTable of
-            Nothing   -> pure Map.empty
-            Just toml -> fmap Map.fromList $ forM (HashMap.toList (tomlPairs toml)) $ \(tomlKey, anyVal) ->
-                case backward valBiMap anyVal of
-                    Right v  -> case forward keyBiMap tomlKey of
-                        Right k  -> pure (k, v)
-                        Left err -> throwError $ BiMapError err
-                    Left err -> throwError $ BiMapError err
+      mTable <- asks $ Prefix.lookup key . tomlTables
+      case mTable of
+        Nothing   -> pure Map.empty
+        Just toml -> fmap Map.fromList $ forM (HashMap.toList (tomlPairs toml)) $ \(tomlKey, anyVal) ->
+          case backward valBiMap anyVal of
+            Right v  -> case forward keyBiMap tomlKey of
+              Right k  -> pure (k, v)
+              Left err -> throwError $ BiMapError err
+            Left err -> throwError $ BiMapError err
+
+    output :: Map key val -> TomlState (Map key val)
+    output dict = do
+      newToml <- foldM update mempty (Map.toList dict)
+      dict <$ modify (insertTable key newToml)
+
+    update :: TOML -> (key, val) -> TomlState TOML
+    update toml (k, v) = do
+      tomlKey <- MaybeT $ pure $ either (const Nothing) Just $ backward keyBiMap k
+      anyVal <- MaybeT $ pure $ either (const Nothing) Just $ forward valBiMap v
+      return $ insertKeyAnyVal tomlKey anyVal toml
+
+dynamicMapCodec :: forall key val .
+                   Ord key
+                => TomlBiMap Key key
+                -> (Key -> TomlCodec val)
+                -> Key
+                -> TomlCodec (Map key val)
+dynamicMapCodec keyBiMap valCodec key = Codec input output
+  where
+    input :: TomlEnv (Map key val)
+    input = do
+      mtable <- asks $ Prefix.lookup key . tomlTables
+      case mtable of
+        Nothing   -> pure Map.empty
+        Just toml -> do
+          simpleVals <- fmap Map.fromList $ forM (HashMap.toList (tomlPairs toml)) $ \(thisKey, anyVal) ->
+            let thisToml = TOML {tomlPairs = HashMap.singleton thisKey anyVal, tomlTables = mempty, tomlTableArrays = mempty}
+                thisCodec = valCodec thisKey in
+              case forward keyBiMap thisKey of
+                Right k  -> case runTomlCodec thisCodec thisToml of
```

Comment:
Converting each `key-value` pair to a separate TOML and running the codec each time seems a bit suboptimal. Need to think about how to do this properly :thinking:

Reply:
Agree that this solution is a no-go at the moment.
@everythingfunctional why exactly `runTomlCodec` is needed here? Seems like a hack, probably there are already functions that are required here in `tomland`?